### PR TITLE
8288752: Split thread implementation files

### PR DESCRIPTION
**Pull request by coleenp**
This changes splits thread.hpp/cpp into javaThread, and threads files.

I left the commits intact to see better the progression of changes, but most files are include file changes.  The only tricky parts are that some files were included in thread.hpp, like mutexLocker.cpp, which has to be included in the files that need it.

I didn't update the copyrights to save diffs but will before integration.
Also I won't integrate until after Dan's JDK 19 changes are merged into JDK 20.

Tested with tier1-4 on Oracle supported platforms and built on other platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288752](https://bugs.openjdk.org/browse/JDK-8288752): Split thread implementation files

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [ae864661](https://git.openjdk.org/jdk/pull/9218/files/ae8646615e0bf76770e4a778b9f8a5f1f2b5c45f)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [93cff18d](https://git.openjdk.org/jdk/pull/9218/files/93cff18da44c191b2de2bf7cd172faff3606cff7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9218/head:pull/9218` \
`$ git checkout pull/9218`

Update a local copy of the PR: \
`$ git checkout pull/9218` \
`$ git pull https://git.openjdk.org/jdk pull/9218/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9218`

View PR using the GUI difftool: \
`$ git pr show -t 9218`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9218.diff">https://git.openjdk.org/jdk/pull/9218.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Hi Coleen,

The changes in themselves seem fine to me - I only skimmed things, mainly looking at the new files.  It is a very large refactoring so there will be an impact on backports naturally.

I'm not sure this will save much in terms of build time inclusions (how many files only include thread.hpp now?) but the code structure/management benefits of this are more important anyway. (Though I must confess in a remote terminal using vi I appreciate everything being in one file. :) ).

Thanks.

**Review by robehn — APPROVED** [review]
Looks good, thanks.

**Comment by coleenp** [comment]
Thanks David and Robbin.
Unfortunately, this change will make backporting more difficult if we have backport changes from JDK 20.  I still think it's wroth separating the thread files.  As a lowly 'vi' user, it's really helpful for me to see whether a field or function belongs to JavaThread or Thread without scrolling up and down the file.  I guess that there might be fields in one that really belongs in the other.  This makes it easier, at least for me, to see.

Thank you for the review, Robbin.

I'm still waiting for GHA to pass as I forgot about shenandoah and some of the other platforms build differently.  Also I'm waiting for @dcubed-ojdk changes to JDK 19 to be integrated and merged into JDK 20 first.

**Review by iklam — APPROVED** [review]

**Comment by iklam** [comment]
> Hi Coleen,
> 
> The changes in themselves seem fine to me - I only skimmed things, mainly looking at the new files. It is a very large refactoring so there will be an impact on backports naturally.
> 
> I'm not sure this will save much in terms of build time inclusions (how many files only include thread.hpp now?) but the code structure/management benefits of this are more important anyway. (Though I must confess in a remote terminal using vi I appreciate everything being in one file. :) ).
> 
> Thanks.

Currently, as counted by https://github.com/iklam/tools/blob/main/headers/whoincludes.tcl, there isn't much improvement in terms of build time:

Before this PR:
- thread.hpp was included 847 times

After this PR
- thread.hpp is included 848 times
- javaThread.hpp is included 843 times

One reason is in handles.hpp, which has implicit conversion from `JavaThread*` to `Thread*`. E.g., when you have:

```
void blah(TRAPS) {
    Handle h(THREAD, someOop);
   . ..
}
```

Anyway, I think this PR is a very important step towards faster build time. We can do further refactoring. E.g., avoid including handles.hpp unnecessarily in other .hpp files (and move offenders from .hpp to .inline.hpp, etc).

**Comment by coleenp** [comment]
I took out #include "mutexLocker.hpp" from thread.hpp, so on the minimal variant, this isn't transitively included from some other file.

**Comment by dcubed-ojdk** [comment]
My three fixes have been sync'ed from JDK19 to JDK20.

**Comment by coleenp** [comment]
@dcubed-ojdk Your changes are now in threads.cpp and javaThread.cpp/hpp.  Can you have a look to make sure they're correct?  They were straightforward to move.

**Comment by tstuefe** [comment]
I like this a lot :) Thanks for doing this. I put this into our CI, just to be sure, but since Aleksey added most of our platforms to GHAs I don't expect any surprises.

**Review by dholmes-ora — APPROVED** [review]

File: src/hotspot/share/prims/jvmtiDeferredUpdates.hpp [hunk 1]
```diff
@@ -26,10 +26,13 @@
...
 
-#include "runtime/thread.inline.hpp"
+#include "runtime/javaThread.hpp"
 #include "utilities/globalDefinitions.hpp"
 #include "utilities/growableArray.hpp"
 
+class MonitorInfo;
+template <typename T> class GrowableArray;
+
 class jvmtiDeferredLocalVariable : public CHeapObj<mtCompiler> {
```
Comment (dcubed-ojdk) [inline src/hotspot/share/prims/jvmtiDeferredUpdates.hpp:36]: This surprised me, but I'm guessing we picked up the template
from thread.inline.hpp...

File: src/hotspot/share/runtime/interfaceSupport.inline.hpp [hunk 2]
```diff
@@ -30,12 +30,13 @@
...
 #include "gc/shared/gc_globals.hpp"
 #include "runtime/handles.inline.hpp"
+#include "runtime/javaThread.inline.hpp"
 #include "runtime/mutexLocker.hpp"
 #include "runtime/orderAccess.hpp"
 #include "runtime/os.hpp"
 #include "runtime/safepointMechanism.inline.hpp"
 #include "runtime/safepointVerifiers.hpp"
-#include "runtime/thread.hpp"
+#include "runtime/javaThread.inline.hpp"
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/interfaceSupport.inline.hpp:39]: This is a duplicate include of the one on L33 above.

File: src/hotspot/share/runtime/vframe.hpp [hunk 3]
```diff
@@ -113,6 +113,7 @@ class vframe: public ResourceObj {
 #endif
 };
 
+class MonitorInfo;
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/vframe.hpp:116]: This one also surprise me. I don't have a guess how this one worked before.

**Review by dcubed-ojdk — APPROVED** [review]
Sorry for the late review. I decided to try a crawl-thru review with this PR
and managed to get 1/3 done yesterday and the rest done today.

I very much like this PR. I think there's only one comment that will require a cleanup.

**Comment by dcubed-ojdk** [comment]
@coleenp's comment from above:
> @dcubed-ojdk Your changes are now in threads.cpp and javaThread.cpp/hpp.
> Can you have a look to make sure they're correct? They were straightforward to move.

I double checked each of the three patches and all the changes are there and in
the right places. What was really cool was being able to pick between the javaThread.*,
thread.* and threads.* files easily based on the class name for the function that I
was changing. Pretty easy to navigate the new layout (IMHO).

File: src/hotspot/share/prims/jvmtiDeferredUpdates.hpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/prims/jvmtiDeferredUpdates.hpp:36]: yes.

**Comment by dcubed-ojdk** [comment]
No apology needed. It was a huge change and you don't want to wait too long
and run into merge after merge...

**Pull request closed without merging**